File: spk/api/_install_spec.py
from typing import Dict, List, Any, Iterable
from dataclasses import dataclass, field

from ._request import Request
from ._ident import Ident


@dataclass
class InstallSpec:
    """A set of structured installation parameters for a package."""

    requirements: List[Request] = field(default_factory=list)

    def upsert_requirement(self, request: Request) -> None:
        """Add or update a requirement to the set of installation requirements.

        If a request exists for the same package, it is replaced with the given
        one. Otherwise the new request is appended to the list.
        """
        for i, other in enumerate(self.requirements):
            if other.pkg.name == request.pkg.name:
                self.requirements[i] = request
                return
        else:
            self.requirements.append(request)

    def to_dict(self) -> Dict[str, Any]:
        return {"requirements": list(r.to_dict() for r in self.requirements)}

    def render_all_pins(self, resolved: Iterable[Ident]) -> None:
        """Render all requests with a package pin using the given resolved packages."""

        by_name = dict((pkg.name, pkg) for pkg in resolved)
        for i, request in enumerate(self.requirements):
            if not request.pin:
                continue
            if request.pkg.name not in by_name:
                raise ValueError(f"Pinned package not present: {request.pkg.name}")
            self.requirements[i] = request.render_pin(by_name[request.pkg.name])

    @staticmethod
    def from_dict(data: Dict[str, Any]) -> "InstallSpec":

        spec = InstallSpec()

        requirements = data.pop("requirements", [])
        if requirements:
            spec.requirements = list(Request.from_dict(r) for r in requirements)

        if len(data):
            raise ValueError(
                f"unrecognized fields in spec.install: {', '.join(data.keys())}"
            )

        return spec

File: crates/spk-schema/src/v0/install_spec.rs
// Copyright (c) Sony Pictures Imageworks, et al.
// SPDX-License-Identifier: Apache-2.0
// https://github.com/imageworks/spk
use serde::{Deserialize, Serialize};
use spk_schema_ident::BuildIdent;

use super::EmbeddedPackage;
use crate::foundation::option_map::OptionMap;
use crate::ident::Request;
use crate::{ComponentSpecList, EnvOp, RequirementsList, Result};

#[cfg(test)]
#[path = "./install_spec_test.rs"]
mod install_spec_test;

/// A set of structured installation parameters for a package.
#[derive(Clone, Debug, Default, Deserialize, Eq, Hash, Ord, PartialEq, PartialOrd, Serialize)]
pub struct InstallSpec {
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub requirements: RequirementsList,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub embedded: Vec<EmbeddedPackage>,
    #[serde(default)]
    pub components: ComponentSpecList<EmbeddedPackage>,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub environment: Vec<EnvOp>,
}

impl InstallSpec {
    /// Add or update a requirement to the set of installation requirements.
    ///
    /// If a request exists for the same name, it is replaced with the given
    /// one. Otherwise the new request is appended to the list.
    pub fn upsert_requirement(&mut self, request: Request) {
        self.requirements.upsert(request);
    }

    pub fn is_default(&self) -> bool {
        self.requirements.is_empty() && self.embedded.is_empty() && self.components.is_default()
    }

    /// Render all requests with a package pin using the given resolved packages.
    pub fn render_all_pins<'a>(
        &mut self,
        options: &OptionMap,
        resolved: impl Iterator<Item = &'a BuildIdent>,
    ) -> Result<()> {
        self.requirements.render_all_pins(options, resolved)
    }
}
